# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



# Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

# Входные данные от банка
- статистика о платёжеспособности клиентов.

# Описание данных
- `children` — количество детей в семье
- `days_employed` — общий трудовой стаж в днях
- `dob_years` — возраст клиента в годах
- `education` — уровень образования клиента
- `education_id` — идентификатор уровня образования
- `family_status` — семейное положение
- `family_status_id` — идентификатор семейного положения
- `gender` — пол клиента
- `income_type` — тип занятости
- `debt` — имел ли задолженность по возврату кредитов
- `total_income` — ежемесячный доход
- `purpose` — цель получения кредита


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

data = pd.read_csv('../datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E до 30.000 руб.'
        elif 30000 < income <= 50000:
            return 'D 30.000 - 50.000 руб.'
        elif 50000 < income <= 200000:
            return 'C 50.000 - 200.000 руб.'
        elif 200000 < income <= 1000000:
            return 'B 200.000 - 1.000.000 руб.'
        elif income > 1000000:
            return 'A свыше 1.000.000 руб.'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Создатим новый датафрейм групиировкой по 'children' (количество детей в семье) с функциями 'sum', 'count', 'mean' по 'debt' (имел ли задолженность по возврату кредитов)

data_child_dept = data.groupby('children').agg({'debt':['sum', 'count', 'mean']})
# mean в нашей задаче считает это не только среднее, но и % заемщиков имеющих задолженость, отформатируем его в %
data_child_dept[('debt',  'mean')] = data_child_dept[('debt',  'mean')].transform(lambda x: '{:,.2%}'.format(x))
data_child_dept

debt              
           sum  count   mean
children                    
0         1063  14091  7.54%
1          444   4808  9.23%
2          194   2052  9.45%
3           27    330  8.18%
4            4     41  9.76%
5            0      9  0.00%

**Вывод:** 

Существует очень тонкая связь между количество детей в семье и возвратом кредитов. 
Если принимать только процент заёмщиков имевших задолженность по возврату кредитов, то самая ответственная группа - семьи, имеющие пять детей, а самая "проблемная" группа заемщиков в семьях, имеющих четырех детей. Но по ним очень маленькая выборка, поэтому данные выводы могут нести ошибку. 
Мы можем утверждать, что заемщики с детьми чаще имеют задолженность, чем заемщики без детей.  


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
# Создатим новый датафрейм групиировкой по 'family_status' (семейное положение) с функциями 'sum', 'count', 'mean' по 'debt' (имел ли задолженность по возврату кредитов)

data_family_dept = data.groupby('family_status').agg({'debt':['sum', 'count', 'mean']})
# mean в нашей задаче считает это не только среднее, но и % заемщиков имеющих задолженость, отформатируем его в %
data_family_dept[('debt',  'mean')] = data_family_dept[('debt',  'mean')].transform(lambda x: '{:,.2%}'.format(x))
data_family_dept

debt              
                       sum  count   mean
family_status                           
Не женат / не замужем  273   2796  9.76%
в разводе               84   1189  7.06%
вдовец / вдова          63    951  6.62%
гражданский брак       385   4134  9.31%
женат / замужем        927  12261  7.56%

**Вывод:** 

Самая ответственная группа по возврату кредита в срок **вдовец / вдова** имеет всего 6.62% заемщиков, которые имели задолженность по возврату кредитов. Немного уступает группа **в разводе** имеющая 7.06% заемщиков, которые имели задолженность по возврату кредитов.  Замыкает тройку группа **женат / замужем** с 7.56% заемщиков, которые имели задолженность по возврату кредитов. 
Проблемными группами можем считать **Не женат / не замужем** с 9.76% и **гражданский брак** с 9.31% заемщиков, которые имели задолженность по возврату кредитов.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
# Создатим новый датафрейм групиировкой по 'total_income_category' (категории доходов) с функциями 'sum', 'count', 'mean' по 'debt' (имел ли задолженность по возврату кредитов)

data_income_categ_dept = data.groupby('total_income_category').agg({'debt':['sum', 'count', 'mean']})
# mean в нашей задаче считает это не только среднее, но и % заемщиков имеющих задолженость, отформатируем его в %
data_income_categ_dept[('debt',  'mean')] = data_income_categ_dept[('debt',  'mean')].transform(lambda x: '{:,.2%}'.format(x))
data_income_categ_dept

debt              
                             sum  count   mean
total_income_category                         
A свыше 1.000.000 руб.         2     25  8.00%
B 200.000 - 1.000.000 руб.   354   5014  7.06%
C 50.000 - 200.000 руб.     1353  15921  8.50%
D 30.000 - 50.000 руб.        21    349  6.02%
E до 30.000 руб.               2     22  9.09%

**Вывод:** 

Мы имеем 5 категорий по доходу заемщика, две из которых составляют более 98% всей выборки: 
- **B 200.000 - 1.000.000 руб.**
- **C 50.000 - 200.000 руб.**

Из данных групп низкий процент заемщиков, которые имели задолженность по возврату кредитов, у категории **B 200.000 - 1.000.000 руб.** с показателем 7.06%, что ниже на 1.44 пунктов (составляет 20% роста) в категории **C 50.000 - 200.000 руб.** с показателем 8.50%. Скорее всего, заемщики с более высоким доходом подходят обдуманнее к размеру кредита и возможности его вернуть. 

Малочисленные категории составляют менее 2% всей выборки: 
- **A свыше 1.000.000 руб.**
- **D 30.000 - 50.000 руб.**
- **E до 30.000 руб.**	

Анализ малочисленных категорий может привести к ошибочному результату.


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
# Создатим новый датафрейм групиировкой по 'purpose_category' (категории целей кредита) с функциями 'sum', 'count', 'mean' по 'debt' (имел ли задолженность по возврату кредитов)

data_purpose_category_dept = data.groupby('purpose_category').agg({'debt':['sum', 'count', 'mean']})
# mean в нашей задаче считает это не только среднее, но и % заемщиков имеющих задолженость, отформатируем его в %
data_purpose_category_dept[('debt',  'mean')] = data_purpose_category_dept[('debt',  'mean')].transform(lambda x: '{:,.2%}'.format(x))
data_purpose_category_dept

debt              
                          sum  count   mean
purpose_category                           
операции с автомобилем    400   4279  9.35%
операции с недвижимостью  780  10751  7.26%
получение образования     369   3988  9.25%
проведение свадьбы        183   2313  7.91%

**Вывод:** 

Группа **операции с недвижимостью** явлется лучшей по возрату кредита в срок с показателем 7.26%. Также хорошо возращают кредиты с группы **проведение свадьбы** с показателем 7.91%. 
Группы **операции с автомобилем** и **получение образования** имеют выше процент возрата кредита в долг: 9.35% и 9.25% соответственно. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*

Пропуски в столбцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход) совпадают, возможные причины:
- при заполнении анкеты не указал заемщик days_employed (общий трудовой стаж в днях) и ему не открылось поле total_income (ежемесячный доход), возможно, была реализована логика: нет стажа - нет дохода.
- по техническим причинам: при обмене данных с филиалами не прошли данные по данным столбцам
- из-за ошибок или неверности заполнения данных столбцов: days_employed == 0 & total_income > 0, т.к. не может дохода без стажа и данные по столбцам обнулялись. 
- нет и стажа и дохода, поэтому данные пункты пропустил заемщик,
- была акция предоставление займа без подтверждения стажа и дохода. 


#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медианное значение более точно передает среднюю величину выборки, особенно где некторые данные сильно отличаються от большиства. 
Для примера, есть 5 человек, которые едят капусту, и к ним присоединился 6 человек, который ест мясо. В среднем, они едят голубцы, но реально большинство ест капусту, что и отразит медиана.

### Шаг 4: общий вывод.

# В ходе работы были добавлены данные:
- `total_income_category` — категории по ежемесячному доходу 	
- `purpose_category` — категории по целям получения кредита

# При предобработке данных появились вопросы для уточнения:
1. аномальные значения по столбцу `children` (количество детей в семье): -1 и 20. Можно убрать или -1 — это ошибка, а 20 многодетный родитель? В данном проекте строки с аномальными значениями удалены; 
2. известны ли причины пропусков в столбцах `days_employed` (общий трудовой стаж в днях) и `total_income` (ежемесячный доход)? в данном проекте пропуски заменены на медианные значения данных столбцов; 
3. в столбце `days_employed` есть аномально большие значения (стаж работы/365 больше возраста) и почему они дробные, неужели считали до секунды стаж. Столбец без исправлений, т.к. он не понадобится для исследования;
4. в столбце `dob_years` (возраст клиента в годах) есть нулевые значения, что не может быть. Столбец без исправлений, т.к. он не понадобится для исследования;
5. в столбце `gender` (пол клиента) есть XNA, это ошибка или нет? Столбец без исправлений, т.к. он не понадобится для исследования;

# Выводы 
1. Между *количеством* детей и возвратом кредита в срок прямой зависимости нет, но вот зависимость между наличием детей и возвратом кредита в срок есть. Категория **заемщиков без детей** имеет всего 7,54% заемщиков, которые имели задолженность по возврату кредитов, когда в категории **заемщиков с 1 ребенком** таких заемщиков уже 9,23%, что выше на 1,69 пункт или на 22%. Не рассматривал категории **заемщиков от 3 детей в семье** ввиду их малочисленности. Возможно, задержки с выплатами в срок в семьях с детьми чаще происходят из-за непредвиденных расходов, связанные с детьми, и любовь к детям перевешивает ответственность перед банком. 
    
2. Между семейным положением и возвратом кредита в срок самые ответственные группы **вдовец / вдова** с 6.62% заемщиков, которые имели задолженность по возврату кредитов, **в разводе** с 7.06% и **женат / замужем** с 7.56%. Сильно отстают группы **гражданский брак** с 9.31% и **Не женат / не замужем** с 9.76%. Разрыв между третьей группой в рейтинге и четвертой составляет 1.75 пункта или 23%. Возможно, что первые группы более ответственные, т.к. они уже брали на себя ответственность при вступлении в брак, а вторая пока с такой ответственностью еще не знакома.
    
3. Между уровнем дохода и возвратом кредита в срок мы имеем 5 категорий по доходу заемщика, две из которых составляют более 98% всей выборки: **B 200.000 - 1.000.000 руб.** и **C 50.000 - 200.000 руб.**. Из данных групп низкий процент заемщиков, которые имели задолженность по возврату кредитов, у категории **B 200.000 - 1.000.000 руб.** с показателем 7.06%, что ниже на 1.44 пунктов (составляет 20% роста) в категории **C 50.000 - 200.000 руб.** с показателем 8.50%. Скорее всего, заемщики с более высоким доходом подходят обдуманнее к размеру и возврату кредита. Малочисленные категории составляют менее 2% всей выборки: **A свыше 1.000.000 руб.**, **D 30.000 - 50.000 руб.** и **E до 30.000 руб.**. Анализ данных категорий не сделан, т.к. он может привести к ошибочному результату.
    
4. Цели кредита влияют на его возврат в срок, так группа **операции с недвижимостью** является лучшей по возврату кредита в срок с показателем 7.26% заемщиков, которые имели задолженность по возврату кредитов. Также хорошо возвращают кредиты с группы **проведение свадьбы** с показателем 7.91%. Группы **операции с автомобилем** и **получение образования** имеют выше процент возврата кредита в долг: 9.35% и 9.25% соответственно. Между второй и третьей группой разница в 1.43 пункта или 17% рост процента заемщиков с долгами. Возможно, что свадьбу покрывают за счет свадебных подарков, а образование принесет бонусы только после окончания образование, да и то не сразу. А отношение к недвижимости более ответственное, чем к автомобилю, т.к. недвижимость важнее для человека. 



